### ISSUE TITLE
hip cannot be concretized due to an invalid variant value error

### ISSUE THREAD

**Issue by haampie** [LLNL/spack#19951]
### Steps to reproduce the issue

```console
$ spack spec hip
==> Error: 'none' cannot be combined with other variant values.
```

but using the latest version explicitly:

```console
$ spack spec hip@3.9.0
```
is fine.

From `spack --debug spec hip` it looks like the issue arises due to this line:

```python
    depends_on('mesa~llvm@18.3: swr=none', type='link')
```
in hip-rocclr: https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/hip-rocclr/package.py#L33

When printing the output variant.py > _value_setter, this is what I see:
```
==> [2020-11-17-09:30:55.173335] hip-rocclr applying constraint mesa@18.3:~llvm swr=none
['auto', 'none']
```
the `auto` is the defautl value, `none` is set by hip.

### Information on your system

* **Spack:** 0.15.4-1979-423e80af2
* **Python:** 3.8.5
* **Platform:** linux-ubuntu20.04-zen2

### Additional information

- [x] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [x] I have searched the issues of this repo and believe this is not a duplicate
- [x] I have run the failing commands in debug mode and reported the output

**Comment by alalazo** [comment]
@haampie This is a concretization issue that should be solved by the new concretizer. It is yet another instance of the original concretizer being greedy and fixing details before having accounted for the whole set of constraints. 

Reporting below the results I obtain with the original concretizer:
```console
$ spack spec hip
Input spec
--------------------------------
hip

Concretized
--------------------------------
==> Error: 'none' cannot be combined with other variant values.
```
and with the new concretizer (that solves the full set of constraints):
```console
$ spack solve hip
==> Best of 0 answers.
==> Optimization: [0, 0, -2, 130, -1, 2, -84, 1, 0, -72, 0]
hip@3.9.0%gcc@10.1.0~ipo build_type=RelWithDebInfo patches=2a4190477b7d9206b9cd8d70770ba0bc007273cbe54772efb12f9ca2e37c0392,a6dbc497d8195e702d989b481f6dc2cc3a61ee9c7952d584e9acc38d2ec4e24c arch=linux-ubuntu18.04-broadwell
    ^cmake@3.18.4%gcc@10.1.0~doc+ncurses+openssl+ownlibs~qt patches=bf695e3febb222da2ed94b3beea600650e4318975da90e4a71d6f31a6d5d8c3d arch=linux-ubuntu18.04-broadwell
    ^comgr@3.9.0%gcc@10.1.0~ipo build_type=Release arch=linux-ubuntu18.04-broadwell
        ^llvm-amdgpu@3.9.0%gcc@10.1.0~ipo build_type=Release arch=linux-ubuntu18.04-broadwell
            ^ncurses@6.2%gcc@10.1.0~symlinks+termlib arch=linux-ubuntu18.04-broadwell
                ^pkgconf@1.7.3%gcc@10.1.0 arch=linux-ubuntu18.04-broadwell
            ^python@3.8.6%gcc@10.1.0+bz2+ctypes+dbm~debug+libxml2+lzma~nis~optimizations+pic+pyexpat+pythoncmd+readline+shared+sqlite3+ssl~tix~tkinter~ucs4+uuid+zlib patches=0d98e93189bc278fbc37a50ed7f183bd8aaf249a8e1670a465f0db6bb4f8cf87 arch=linux-ubuntu18.04-broadwell
                ^bzip2@1.0.8%gcc@10.1.0+shared arch=linux-ubuntu18.04-broadwell
                    ^diffutils@3.7%gcc@10.1.0 arch=linux-ubuntu18.04-broadwell
                        ^libiconv@1.16%gcc@10.1.0 arch=linux-ubuntu18.04-broadwell
                ^expat@2.2.10%gcc@10.1.0+libbsd arch=linux-ubuntu18.04-broadwell
                    ^libbsd@0.10.0%gcc@10.1.0 arch=linux-ubuntu18.04-broadwell
                ^gdbm@1.18.1%gcc@10.1.0 patches=7f569b9287e9e5a4236636eef32ff3428f0462ec17aea359e0ebee07649f1f16 arch=linux-ubuntu18.04-broadwell
                    ^readline@8.0%gcc@10.1.0 arch=linux-ubuntu18.04-broadwell
                ^gettext@0.21%gcc@10.1.0+bzip2+curses+git~libunistring+libxml2+tar+xz arch=linux-ubuntu18.04-broadwell
                    ^libxml2@2.9.10%gcc@10.1.0~python arch=linux-ubuntu18.04-broadwell
                        ^xz@5.2.5%gcc@10.1.0~pic arch=linux-ubuntu18.04-broadwell
                        ^zlib@1.2.11%gcc@10.1.0+optimize+pic+shared arch=linux-ubuntu18.04-broadwell
                    ^tar@1.32%gcc@10.1.0 arch=linux-ubuntu18.04-broadwell
                ^libffi@3.3%gcc@10.1.0 patches=26f26c6f29a7ce9bf370ad3ab2610f99365b4bdd7b82e7c31df41a3370d685c0 arch=linux-ubuntu18.04-broadwell
                ^libuuid@1.0.3%gcc@10.1.0 arch=linux-ubuntu18.04-broadwell
                ^openssl@1.1.1h%gcc@10.1.0+systemcerts arch=linux-ubuntu18.04-broadwell
                    ^perl@5.32.0%gcc@10.1.0+cpanm+shared+threads arch=linux-ubuntu18.04-broadwell
                        ^berkeley-db@18.1.40%gcc@10.1.0 arch=linux-ubuntu18.04-broadwell
                ^sqlite@3.33.0%gcc@10.1.0+column_metadata+fts~functions~rtree arch=linux-ubuntu18.04-broadwell
            ^z3@4.8.7%gcc@10.1.0+python arch=linux-ubuntu18.04-broadwell
                ^py-setuptools@50.3.2%gcc@10.1.0 arch=linux-ubuntu18.04-broadwell
        ^rocm-cmake@3.9.0%gcc@10.1.0~ipo build_type=Release arch=linux-ubuntu18.04-broadwell
        ^rocm-device-libs@3.9.0%gcc@10.1.0~ipo build_type=Release arch=linux-ubuntu18.04-broadwell
            ^texinfo@6.5%gcc@10.1.0 patches=12f6edb0c6b270b8c8dba2ce17998c580db01182d871ee32b7b6e4129bd1d23a,1732115f651cff98989cb0215d8f64da5e0f7911ebf0c13b064920f088f2ffe1 arch=linux-ubuntu18.04-broadwell
    ^hip-rocclr@3.9.0%gcc@10.1.0~ipo build_type=RelWithDebInfo arch=linux-ubuntu18.04-broadwell
        ^hsa-rocr-dev@3.9.0%gcc@10.1.0~ipo build_type=Release patches=71e6851d9be8113bfb8d71b66a3171e0d0401bae5e6f161c9e7fe32558261a46 arch=linux-ubuntu18.04-broadwell
            ^hsakmt-roct@3.9.0%gcc@10.1.0~ipo build_type=Release arch=linux-ubuntu18.04-broadwell
                ^numactl@2.0.14%gcc@10.1.0 patches=4e1d78cbbb85de625bad28705e748856033eaafab92a66dffd383a3d7e00cc94 arch=linux-ubuntu18.04-broadwell
                    ^autoconf@2.69%gcc@10.1.0 arch=linux-ubuntu18.04-broadwell
                        ^m4@1.4.18%gcc@10.1.0+sigsegv patches=3877ab548f88597ab2327a2230ee048d2d07ace1062efe81fc92e91b7f39cd00,fc9b61654a3ba1a8d6cd78ce087e7c96366c290bc8d2c299f09828d793b853c8 arch=linux-ubuntu18.04-broadwell
                            ^libsigsegv@2.12%gcc@10.1.0 arch=linux-ubuntu18.04-broadwell
                    ^automake@1.16.2%gcc@10.1.0 arch=linux-ubuntu18.04-broadwell
                    ^libtool@2.4.6%gcc@10.1.0 arch=linux-ubuntu18.04-broadwell
            ^libelf@0.8.13%gcc@10.1.0 arch=linux-ubuntu18.04-broadwell
        ^mesa@20.2.1%gcc@10.1.0+glx~llvm+opengl~opengles+osmesa buildtype=release swr=none arch=linux-ubuntu18.04-broadwell
            ^binutils@2.35.1%gcc@10.1.0+gold~headers~interwork~ld~libiberty~lto+nls~plugins arch=linux-ubuntu18.04-broadwell
            ^bison@3.6.4%gcc@10.1.0 arch=linux-ubuntu18.04-broadwell
                ^help2man@1.47.11%gcc@10.1.0 arch=linux-ubuntu18.04-broadwell
            ^flex@2.6.4%gcc@10.1.0+lex patches=09c22e5c6fef327d3e48eb23f0d610dcd3a35ab9207f12e0f875701c677978d3 arch=linux-ubuntu18.04-broadwell
                ^findutils@4.6.0%gcc@10.1.0 patches=84b916c0bf8c51b7e7b28417692f0ad3e7030d1f3c248ba77c42ede5c1c5d11e,bd9e4e5cc280f9753ae14956c4e4aa17fe7a210f55dd6c84aa60b12d106d47a2 arch=linux-ubuntu18.04-broadwell
            ^glproto@1.4.17%gcc@10.1.0 arch=linux-ubuntu18.04-broadwell
                ^util-macros@1.19.1%gcc@10.1.0 arch=linux-ubuntu18.04-broadwell
            ^libx11@1.6.7%gcc@10.1.0 arch=linux-ubuntu18.04-broadwell
                ^inputproto@2.3.2%gcc@10.1.0 arch=linux-ubuntu18.04-broadwell
                ^kbproto@1.0.7%gcc@10.1.0 arch=linux-ubuntu18.04-broadwell
                ^libxcb@1.13%gcc@10.1.0 arch=linux-ubuntu18.04-broadwell
                    ^libpthread-stubs@0.4%gcc@10.1.0 arch=linux-ubuntu18.04-broadwell
                    ^libxau@1.0.8%gcc@10.1.0 arch=linux-ubuntu18.04-broadwell
                        ^xproto@7.0.31%gcc@10.1.0 arch=linux-ubuntu18.04-broadwell
                    ^libxdmcp@1.1.2%gcc@10.1.0 arch=linux-ubuntu18.04-broadwell
                    ^xcb-proto@1.13%gcc@10.1.0 arch=linux-ubuntu18.04-broadwell
                ^xextproto@7.3.0%gcc@10.1.0 arch=linux-ubuntu18.04-broadwell
                ^xtrans@1.3.5%gcc@10.1.0 arch=linux-ubuntu18.04-broadwell
            ^libxext@1.3.3%gcc@10.1.0 arch=linux-ubuntu18.04-broadwell
            ^libxt@1.1.5%gcc@10.1.0 arch=linux-ubuntu18.04-broadwell
                ^libice@1.0.9%gcc@10.1.0 arch=linux-ubuntu18.04-broadwell
                ^libsm@1.2.3%gcc@10.1.0 arch=linux-ubuntu18.04-broadwell
            ^meson@0.56.0%gcc@10.1.0 patches=7a7ff3bffb5c1996faf178da6765d578394624edb96fe9297bf21f2e3cd9ebda arch=linux-ubuntu18.04-broadwell
                ^ninja@1.10.1%gcc@10.1.0 arch=linux-ubuntu18.04-broadwell
            ^py-mako@1.0.4%gcc@10.1.0 arch=linux-ubuntu18.04-broadwell
                ^py-markupsafe@1.1.1%gcc@10.1.0 arch=linux-ubuntu18.04-broadwell
            ^xrandr@1.5.0%gcc@10.1.0 arch=linux-ubuntu18.04-broadwell
                ^libxrandr@1.5.0%gcc@10.1.0 arch=linux-ubuntu18.04-broadwell
                    ^libxrender@0.9.10%gcc@10.1.0 arch=linux-ubuntu18.04-broadwell
                        ^renderproto@0.11.1%gcc@10.1.0 arch=linux-ubuntu18.04-broadwell
                    ^randrproto@1.5.0%gcc@10.1.0 arch=linux-ubuntu18.04-broadwell
    ^rocminfo@3.9.0%gcc@10.1.0~ipo build_type=RelWithDebInfo arch=linux-ubuntu18.04-broadwell
```

**Comment by alalazo** [comment]
If the spec seems correct I would close the issue since #19501 seems to handle `swr=none` just fine.

**Comment by haampie** [comment]
Thanks for pointing that out. Happy to see the new concretizer is smarter.

Would it still be OK to patch the old concretizer with #19952? There's already sufficiently many issues with the ROCm packages, and adding the new concretizer as a requirement which is still considered experimental is yet another hurdle to cross.

**Comment by haampie** [comment]
Or maybe we can just drop the variant in hip-rocclr if it isn't vital. @arjun-raj-kuppala and @gkathirv, is https://github.com/spack/spack/blob/develop/var/spack/repos/builtin/packages/hip-rocclr/package.py#L33 a strict requirement, or can we drop the `swr=none` bit to overcome the above current-default-spack-concretizer issue?

**Comment by alalazo** [comment]
> Would it still be OK to patch the old concretizer with #19952?

No, the patch there seems to be wrong. It doesn't act on the concretizer directly, but on a method of an object used  by the concretizer and modifies it in a way that appears to be semantically incorrect. The problem is that the modification doesn't constrain a variant, it just overwrites it if you pass `none` as a value.

**Comment by alalazo** [comment]
Closing as fixed by #19501
